### PR TITLE
Variable node pricing

### PR DESCRIPTION
**Pull request by nicktimko**
Still needs testing, but I reordered some events when acquiring the reservation; I'm not sure if there's a sensitivity I'm not fully aware of that lead to breaking up the `if usage_enforcement:` blocks.

### CODE REVIEW

**Comment by nicktimko** [comment]
Log from testing:

* Create 8 day lease with compute (no rate defined --> 1 SU/hr) node
  * 2017-05-23 16:23:19.946 98354 INFO climate.plugins.oshosts.host_plugin [-] Increasing encumbered for project FG-392 by 191.98 (191.98 hours @ 1.00 SU/hr)
  * 2017-05-23 16:23:19.947 98354 INFO climate.plugins.oshosts.host_plugin [-] Usage encumbered for project FG-392 now 10648.49000000000000021
  * 2017-05-23 16:23:19.947 98354 INFO climate.plugins.oshosts.host_plugin [-] Removing lease exception for user None
* Prolong by "7d"
  * 2017-05-23 16:23:45.222 98354 INFO climate.plugins.oshosts.host_plugin [-] Increasing encumbered for project FG-392 by 168.00 (168.00 hours @ 1.00 SU/hr)
  * 2017-05-23 16:23:45.223 98354 INFO climate.plugins.oshosts.host_plugin [-] Usage encumbered for project FG-392 now 10816.49000000000000021
* Reduce by "10d"
  * 2017-05-23 16:23:55.207 98354 INFO climate.plugins.oshosts.host_plugin [-] Increasing encumbered for project FG-392 by -240.00 (-240.00 hours @ 1.00 SU/hr)
  * 2017-05-23 16:23:55.208 98354 INFO climate.plugins.oshosts.host_plugin [-] Usage encumbered for project FG-392 now 10576.49000000000000021
* Delete lease
  * 2017-05-23 16:24:00.459 98354 INFO climate.plugins.oshosts.host_plugin [-] Decreasing encumbered for project FG-392 by 119.98 (-119.98 hours @ 1.00 SU/hr)
  * 2017-05-23 16:24:00.460 98354 INFO climate.plugins.oshosts.host_plugin [-] Usage encumbered for project FG-392 now 10456.50666666700000018

* Create 1 day lease with m40 (44.4 SU/hr) node
  * 2017-05-23 16:24:15.318 98354 INFO climate.plugins.oshosts.host_plugin [-] Increasing encumbered for project FG-392 by 1064.86 (23.98 hours @ 44.40 SU/hr)
  * 2017-05-23 16:24:15.319 98354 INFO climate.plugins.oshosts.host_plugin [-] Usage encumbered for project FG-392 now 11521.3666666670000005
  * 2017-05-23 16:24:15.319 98354 INFO climate.plugins.oshosts.host_plugin [-] Removing lease exception for user None
* Prolong by "7d"
  * 2017-05-23 16:24:29.179 98354 INFO climate.plugins.oshosts.host_plugin [-] Increasing encumbered for project FG-392 by 7459.20 (168.00 hours @ 44.40 SU/hr)
  * 2017-05-23 16:24:29.180 98354 INFO climate.plugins.oshosts.host_plugin [-] Usage encumbered for project FG-392 now 18980.56666666699999979
* Prolong by "7d"
  * 2017-05-23 16:24:41.747 98354 TRACE oslo_messaging.rpc.dispatcher   File "/usr/lib/python2.7/site-packages/climate/plugins/oshosts/host_plugin.py", line 215, in update_reservation
  * 2017-05-23 16:24:41.750 98354 ERROR oslo_messaging._drivers.common [-] ['Traceback (most recent call last):\n', '  File "...climate/manager/service.py", line 490, in update_lease\n    project_name=project_name)\n', '  File "...climate/plugins/oshosts/host_plugin.py", line 215, in update_reservation\n    \'Update reservation would spend %f more SUs, only %f left\' % (requested, left))\n', 'NotAuthorized: Update reservation would spend 7459.200000 more SUs, only 1019.433333 left\n']
  * **This gives a "Danger: There was an error submitting the form" message in Horizon, rather than a useful message.**
* Delete lease
  * 2017-05-23 16:25:03.227 98354 INFO climate.plugins.oshosts.host_plugin [-] Decreasing encumbered for project FG-392 by 8524.06 (-191.98 hours @ 44.40 SU/hr)
  * 2017-05-23 16:25:03.227 98354 INFO climate.plugins.oshosts.host_plugin [-] Usage encumbered for project FG-392 now 10456.50666666700000018
```

File: climate/plugins/oshosts/host_plugin.py [hunk 1]
```diff
@@ -124,52 +126,57 @@ def create_reservation(self, values, usage_enforcement=False, usage_db_host=None
...
-        )
-        if not host_ids:
-            pool.delete(pool_name)
-            raise manager_ex.NotEnoughHostsAvailable()
-        for host_id in host_ids:
-            db_api.host_allocation_create({'compute_host_id': host_id,
-                                          'reservation_id': reservation['id']})
-
-        if usage_enforcement:
             try:
```
Comment (priteau) [inline climate/plugins/oshosts/host_plugin.py]: If your testing shows it's fine to move the block calling host_reservation_create higher up as you did, you can merge the two try/except blocks interacting with Redis into a single one.

File: climate/plugins/oshosts/host_plugin.py [hunk 2]
```diff
@@ -183,18 +190,25 @@ def update_reservation(self, reservation_id, values, usage_enforcement=False, us
         hosts_in_pool = pool.get_computehosts(
             reservation['resource_id'])
 
+        host_allocations = db_api.host_allocation_get_all_by_values(reservation_id=reservation_id)
+
+
         # Check if we have enough available SUs for update
         if usage_enforcement:
+            total_su_factor = sum(
```
Comment (priteau) [inline climate/plugins/oshosts/host_plugin.py]: The update_reservation function can change the set of reserved hosts, so there is a possibility that for leases without node_type (created with the CLI), the newly allocated hosts may have a different su_factor. We should sum up the total_su_factor after the new matching hosts have been selected, but before the host allocations have been updated.

File: climate/plugins/oshosts/host_plugin.py [hunk 3]
```diff
@@ -36,6 +36,8 @@
 from climate.utils.openstack import nova
 from climate.utils import trusts
 
+from . import billrate
```
Comment (nicktimko) [inline climate/plugins/oshosts/host_plugin.py]: No good reason, fixed.

File: climate/plugins/oshosts/host_plugin.py (hunk 1 above)
Comment (nicktimko) [inline climate/plugins/oshosts/host_plugin.py]: I didn't get any problems, but I'm not sure what to test for it. The name assignment for `host_reservation = host_reservation_create(...)` is actually pointless, and as for side-effects, nothing interacts with it in the allocation calculations, so I'm assuming so long as I preserve order of `host_reservation_create()`, `_matching_hosts()`, and `host_allocation_create`, everything will be fine. The major change is that it is created before checking allocations, so I'm relying on it being backed out cleanly by the caller.

File: climate/plugins/oshosts/host_plugin.py (hunk 2 above)
Comment (nicktimko) [inline climate/plugins/oshosts/host_plugin.py]: If the hosts change (I thought we couldn't do this?) then there'd need to be a backout step where encumbered SUs are released for whatever the remainder of the reservation was (`release(old_total_su_factor * (end_date - now))`), then recomputed for what it is going forward (`encumber(new_total_su_factor * (end_date - now))`).

**Comment by nicktimko** [comment]
Testing with pending commit. My lease creator has no restriction over node_type, so it grabs a mix with various su_factors, and also schedules a collision to force adjustment of the host allocations to test factor recalculation.

image

```
>>> lease_a = cclease.Lease(session, nodes=5, start=T[0], end=T[4])
2017-06-06 18:57:33.579 143714 INFO climate.plugins.oshosts.host_plugin [-] Increasing encumbered for project openstack by 774.40 (16.00 hours @ 48.40 SU/hr)
2017-06-06 18:57:33.644 143714 INFO climate.plugins.oshosts.host_plugin [-] Usage encumbered for project openstack now 774.40000000000000002
2017-06-06 18:57:33.645 143714 INFO climate.plugins.oshosts.host_plugin [-] Removing lease exception for user priteau

>>> lease_b = cclease.Lease(session, nodes=6, start=T[0], end=T[1])
2017-06-06 18:57:37.110 143714 INFO climate.plugins.oshosts.host_plugin [-] Increasing encumbered for project openstack by 3575.20 (4.00 hours @ 893.80 SU/hr)
2017-06-06 18:57:37.176 143714 INFO climate.plugins.oshosts.host_plugin [-] Usage encumbered for project openstack now 4349.60000000000000009
2017-06-06 18:57:37.177 143714 INFO climate.plugins.oshosts.host_plugin [-] Removing lease exception for user priteau

>>> lease_c = cclease.Lease(session, nodes=6, start=T[2], end=T[3])
2017-06-06 18:57:44.986 143714 INFO climate.plugins.oshosts.host_plugin [-] Increasing encumbered for project openstack by 3575.20 (4.00 hours @ 893.80 SU/hr)
2017-06-06 18:57:45.051 143714 INFO climate.plugins.oshosts.host_plugin [-] Usage encumbered for project openstack now 7924.79999999999999982
2017-06-06 18:57:45.052 143714 INFO climate.plugins.oshosts.host_plugin [-] Removing lease exception for user priteau

>>> lease_a.delete()
2017-06-06 18:57:46.795 143714 INFO climate.plugins.oshosts.host_plugin [-] Decreasing encumbered for project openstack by 774.40 (-16.00 hours @ 48.40 SU/hr)
2017-06-06 18:57:46.860 143714 INFO climate.plugins.oshosts.host_plugin [-] Usage encumbered for project openstack now 7150.39999999999999991

```
image

```
>>> lease_b.blazar.lease.update(lease_b.id, prolong_for='12h')
2017-06-06 19:01:46.250 143714 DEBUG climate.plugins.oshosts.host_plugin [-] Dropping host 80 from reservation 9043ced2-a48d-45a5-8a89-249697911ea7 update_reservation /usr/lib/python2.7/site-packages/climate/plugins/oshosts/host_plugin.py:263
2017-06-06 19:01:46.311 143714 DEBUG climate.plugins.oshosts.host_plugin [-] Dropping host 91 from reservation 9043ced2-a48d-45a5-8a89-249697911ea7 update_reservation /usr/lib/python2.7/site-packages/climate/plugins/oshosts/host_plugin.py:263
2017-06-06 19:01:46.336 143714 DEBUG climate.plugins.oshosts.host_plugin [-] Dropping host 81 from reservation 9043ced2-a48d-45a5-8a89-249697911ea7 update_reservation /usr/lib/python2.7/site-packages/climate/plugins/oshosts/host_plugin.py:263
2017-06-06 19:01:46.360 143714 DEBUG climate.plugins.oshosts.host_plugin [-] Dropping host 82 from reservation 9043ced2-a48d-45a5-8a89-249697911ea7 update_reservation /usr/lib/python2.7/site-packages/climate/plugins/oshosts/host_plugin.py:263
2017-06-06 19:01:46.386 143714 DEBUG climate.plugins.oshosts.host_plugin [-] Dropping host 84 from reservation 9043ced2-a48d-45a5-8a89-249697911ea7 update_reservation /usr/lib/python2.7/site-packages/climate/plugins/oshosts/host_plugin.py:263
2017-06-06 19:01:46.410 143714 DEBUG climate.plugins.oshosts.host_plugin [-] Dropping host 79 from reservation 9043ced2-a48d-45a5-8a89-249697911ea7 update_reservation /usr/lib/python2.7/site-packages/climate/plugins/oshosts/host_plugin.py:263
2017-06-06 19:01:46.436 143714 DEBUG climate.plugins.oshosts.host_plugin [-] Adding host 15 to reservation 9043ced2-a48d-45a5-8a89-249697911ea7 update_reservation /usr/lib/python2.7/site-packages/climate/plugins/oshosts/host_plugin.py:269
2017-06-06 19:01:46.464 143714 DEBUG climate.plugins.oshosts.host_plugin [-] Adding host 77 to reservation 9043ced2-a48d-45a5-8a89-249697911ea7 update_reservation /usr/lib/python2.7/site-packages/climate/plugins/oshosts/host_plugin.py:269
2017-06-06 19:01:46.489 143714 DEBUG climate.plugins.oshosts.host_plugin [-] Adding host 78 to reservation 9043ced2-a48d-45a5-8a89-249697911ea7 update_reservation /usr/lib/python2.7/site-packages/climate/plugins/oshosts/host_plugin.py:269
2017-06-06 19:01:46.514 143714 DEBUG climate.plugins.oshosts.host_plugin [-] Adding host 83 to reservation 9043ced2-a48d-45a5-8a89-249697911ea7 update_reservation /usr/lib/python2.7/site-packages/climate/plugins/oshosts/host_plugin.py:269
2017-06-06 19:01:47.027 143714 DEBUG climate.plugins.oshosts.host_plugin [-] Adding host 90 to reservation 9043ced2-a48d-45a5-8a89-249697911ea7 update_reservation /usr/lib/python2.7/site-packages/climate/plugins/oshosts/host_plugin.py:269
2017-06-06 19:01:47.059 143714 DEBUG climate.plugins.oshosts.host_plugin [-] Adding host 93 to reservation 9043ced2-a48d-45a5-8a89-249697911ea7 update_reservation /usr/lib/python2.7/site-packages/climate/plugins/oshosts/host_plugin.py:269
2017-06-06 19:01:47.092 143714 WARNING climate.plugins.oshosts.host_plugin [-] SU factor changing from 893.8 to 10048.4
2017-06-06 19:01:47.092 143714 INFO climate.plugins.oshosts.host_plugin [-] Changing encumbered for project openstack by 157199.20 (refunding 16.00 hours @ 893.80 SU/hr, adding 16.00 hours @ 10048.40 SU/hr)
# the first 16 was a printing error, fixed in commit
2017-06-06 19:01:47.159 143714 INFO climate.plugins.oshosts.host_plugin [-] Usage encumbered for project openstack now 164349.60000000000000853
```
image

The overlap is a glitch, https://github.com/ChameleonCloud/python-blazarclient/pull/4 doesn't seem to be live on the test site, but the new rates are calculated appropriately.

**Comment by nicktimko** [comment]
The code assumes that the lease is in the future so won't refund already-encumbered SUs...might need to change the refund/adding to only hours in the future? I also wonder how the usage reporting scripts deal with shifting a reservation to cheaper/costlier nodes...

**Comment by nicktimko** [comment]
The subclassed `ClimateException` doesn't seem to be handled by the manager any better, the message returned in Horizon or to the HTTP client is just "500 Internal Server Error". I don't see how the `NotEnoughHosts` or `MalformedParameter` are intercepted and handled better with the custom message.

**Comment by priteau** [comment]
I rebased the billrate branch on top of stable/liberty.

**Pull request merged**